// tools/linesDbOpr.js

var mongoose = require('mongoose');

var BusAuthSchema = mongoose.Schema({
        lineid : {type : String, default : ''},
        ownerid : {type : String, default : ''},
});


var MongOpr=require("memod/mongooseOpr");


function AuthDbOpr(dbname,modelSchema,modelName){
    MongOpr.call(this,dbname,modelSchema,modelName);//call the super class construct,note:add this 
}
var ut=require('util');
ut.inherits(AuthDbOpr,MongOpr);

AuthDbOpr.prototype.querryTodayAdd=function(cb){
    var today=new Date();
    var ts=today.getFullYear()+"-"+(today.getMonth()+1)+"-"+today.getDate(); 
    mo.getAll({addate:{"$gte":new Date(ts)}},'',function(err,docs){
        var res="";
        for(var i in docs){
            res=(docs[i].addate+"").split(" ")[4]+"-"+docs[i].mobStuctinfo.APP+"-"+docs[i].mobStuctinfo.Product+"-"+docs[i].mobStuctinfo.VERSION+","+res;
        }
        cb("["+res+"]");
    });
}

var mo = new AuthDbOpr("busLineDb",BusAuthSchema,"BusLine");
mo.init();
module.exports = mo;

//tb();

function tb(){
    mo.getAll({},'',function(err,docs){
        for(var i in docs){
            console.log("docs["+i+"]="+docs[i]);
        }
    });
}
function tA(){
    var mo=new MongOpr("tst",AppAuth);
    mo.init();
    mo.add({mobinfo:"name1"});
    mo.getOne({mobinfo:"name1"},'',function(err,doc){
        for(k in doc){
        }
        console.log('ok,name='+doc.mobinfo+",skill="+doc.authtype+",id="+doc.id+",_id="+doc._id);
    });
    mo.getAt(1000,function(err,doc){
        if(doc){
            console.log('ok,getAt='+doc.mobinfo+",skill="+doc.authtype+",id="+doc.id+",_id="+doc._id);
        }else{
            console.log("err,getAt no doc at such index")
        }
    });

    mo.count({mobinfo:"name1"},function(err,cnt){
        console.log("cnt="+cnt);
    });
}
